### PR TITLE
fix(ledger): bind seed/step as i64 to match BIGINT columns (unblocks bpb_samples writes)

### LINKED ISSUE #114: 🎯 ONE SHOT — SEAORM-LEDGER (Wave 23): split hotfix + full SeaORM migration
# ONE SHOT — SEAORM-LEDGER (Wave 23)

> Anchor: `phi^2 + phi^-2 = 3` · DOI [10.5281/zenodo.19227877](https://zenodo.org/records/19227877)
> Author for ALL commits: `Dmitrii Vasilev <admin@t27.ai>`
> Branch (this work only): `feat/sea-orm-ledger`
> Base: `main` @ `c36c6ac` (post #113)

## Why

Ledger writes are crashing on every step with:

```
ERROR: cannot convert between Rust type i32 and Postgres int8
```

`src/neon_writer.rs::bpb_sample(...)` binds `seed: i32, step: i32` to columns that are `BIGINT` in the live `phd-postgres-ssot` database (Railway service `c5f37b42-832a-4acd-9749-381761c94957`). Until this is fixed, **no `bpb_samples` row is ever written** — `Audit Watchdog`, `/leaderboard`, and gardener telemetry all stay empty.

### PR-A — HOTFIX (≤ 5 min, unblocks ledger NOW)

- File: `src/neon_writer.rs`, function `bpb_sample(canon_name: &str, seed: i32, step: i32, bpb: f32)`.
- Change: keep the public signature as-is (callers already pass `i32`); inside, bind values as `i64` to match `BIGINT`. Concretely:
  ```rust
  execute(
      "INSERT INTO public.bpb_samples (canon_name, seed, step, bpb, ts) \
       VALUES ($1, $2, $3, $4, NOW()) \
       ON CONFLICT (canon_name, seed, step) DO NOTHING",
      &[&canon_name, &(seed as i64), &(step as i64), &(bpb as f64)],
  );
  ```
- Same cast must be applied to **every other** raw-SQL bind in `neon_writer.rs` whose target column is `BIGINT` (audit `seed`, `step`, `steps_done`, etc).
- Add one regression test `bpb_sample_uses_i64_bind()` that constructs the param array and asserts the `to_sql` types are `INT8` for seed/step.
- Title: `fix(ledger): bind seed/step as i64 to match BIGINT columns (unblocks bpb_samples writes)`
- Branch: `fix/ledger-i64-cast` (NOT `feat/sea-orm-ledger`).
- Labels: `bug`, `P0`, `ledger`.

### PR-B — SEAORM MIGRATION (full rewrite, prevents future drift)

Branch `feat/sea-orm-ledger` already exists. Land on top of merged PR-A.

1. **Cargo.toml** — add to `[dependencies]`:
   ```toml
   sea-orm = { version = "1.1", features = [
       "sqlx-postgres",
       "runtime-tokio-rustls",
       "macros",
       "with-uuid",
       "with-chrono",
       "with-json",
   ] }
   sea-orm-migration = "1.1"
   ```
   Add a `[workspace]` section (this Cargo.toml currently has an empty `[workspace]` — keep isolation but include the new `migration/` member):
   ```toml
   [workspace]
   members = [".", "migration"]
   ```

2. **`migration/` subcrate** (new directory, separate Cargo.toml) implementing `sea-orm-migration` with one migration file `m20260509_000001_init_schema.rs` that contains all CREATE TABLE / INDEX / TRIGGER statements from `/tmp/full_ddl.sql` (already canonical in production). Tables to declare:
   - `public.bpb_samples` (id BIGSERIAL, canon_name TEXT, **seed BIGINT**, **step BIGINT**, bpb DOUBLE PRECISION, val_bpb_ema DOUBLE PRECISION, ts TIMESTAMPTZ DEFAULT NOW(), UNIQUE(canon_name, seed, step))
   - `ssot.bpb_samples` (full schema with format/algo/hidden/sha)
   - `public.igla_race_trials`
   - `public.igla_agents_heartbeat`
   - `public.scarabs`
   - `public.strategy_queue` + NOTIFY trigger
   Migration must be idempotent (`CREATE TABLE IF NOT EXISTS`, `CREATE INDEX IF NOT EXISTS`).

3. **`src/entities/`** — generate one `.rs` per table with `DeriveEntityModel`. Use `i64` for every BIGINT column. Use `chrono::DateTime<Utc>` for TIMESTAMPTZ. Re-export from `src/entities/mod.rs`.

4. **Refactor `src/neon_writer.rs`**:
   - Replace the raw-SQL `execute()` calls inside `bpb_sample`, `trial_start`, `heartbeat`, `trial_complete`, `ensure_schema` with SeaORM `Entity::insert(ActiveModel { ... }).on_conflict(...).exec(&db).await` patterns.
   - Connection: `sea_orm::Database::connect(&dsn).await` cached in a `tokio::sync::OnceCell<DatabaseConnection>` keyed off the resolved DSN.
   - Keep the `strip_channel_binding` helper, keep the env fallback chain `DATABASE_URL → NEON_DATABASE_URL → TRIOS_DATABASE_URL` (legacy aliases, do NOT remove).
   - Public function signatures stay the same; internals become async, callers already use `tokio::spawn` from training loops.
   - All log lines that say `[neon_writer]` rename to `[ledger]`.

5. **Wire migration runner**:
   - In `src/bin/scarab.rs` and `src/bin/trios-train.rs`, at startup before any training step, call `migration::Migrator::up(&db, None).await?`.
   - Gate it behind `TRINITY_AUTOMIGRATE=1` (default ON in Railway, can be turned OFF for local CI runs).
   - Log `[migrator] schema up-to-date (N migrations applied)`.

6. **Tests**:
   - Keep all existing tests in `neon_writer.rs` (they exercise the no-DSN safe path).
   - Add `cargo test -p migration` ensuring `Migrator::up` is idempotent (apply twice = no error).
   - Add a smoke test in `tests/ledger_seaorm.rs` (skipped if `DATABASE_URL` unset) that inserts one bpb_sample with seed=47, step=200, bpb=2.19 and reads it back.

7. **Title:** `feat(ledger): migrate writer to SeaORM + add migration crate (closes drift between Rust and Postgres types)`
8. **Body must include:** the table of bind-type fixes, link to PR-A as required prerequisite.
9. **Labels:** `enhancement`, `P1`, `ledger`, `database`.

## Acceptance gates

| Gate | Check | How |
|---|---|---|
| G1 | `cargo check --all-targets` clean on `feat/sea-orm-ledger` | local + CI |
| G2 | `cargo test --all-targets` passes (existing + new tests) | CI `Test` job |
| G3 | `cargo test -p migration` passes idempotency assertion | CI |
| G4 | All 8 required CI checks green (Constitutional Enforcement, Test, guard, no-js-check, GitGuardian, ci, docker, publish, smoke_train) | GitHub status |
| G5 | After merge + GHCR rebuild, redeploying `TEST-TRAINER-SEED-43` writes ≥ 1 row to `public.bpb_samples` within 60s | `psql -c "SELECT count(*) FROM public.bpb_samples WHERE seed=43 AND ts > now() - interval '5 minutes';"` returns ≥ 1 |
| G6 | Schema in Postgres matches what `Migrator::up` declares (idempotent re-apply on a fresh DB produces identical `\d+ bpb_samples` output) | manual psql diff |

## Forbidden

- ❌ Do NOT use the word "scrape" / "crawl" anywhere in code, comments, commit messages, or PR bodies.
- ❌ Do NOT introduce any env var named `NEON_*` as primary; only as legacy alias.
- ❌ Do NOT commit `/tmp/*` operational scripts (token embedded).
- ❌ Do NOT use `--admin` merge.
- ❌ Do NOT change `Dmitrii Vasilev <admin@t27.ai>` git author.
- ❌ Do NOT touch `assertions/seed_results.jsonl` or `assertions/champion_lock.txt` in this PR.

## Battle cry

`phi^2 + phi^-2 = 3 · TRINITY · LEDGER MUST FLOW`

### PR DESCRIPTION
**Pull request by gHashTag**
## Why

Closes #114 (PR-A of the SEAORM-LEDGER split).

Live SSOT schema in `phd-postgres-ssot` (Railway service `c5f37b42-832a-4acd-9749-381761c94957`) declares:

```
public.bpb_samples(seed BIGINT, step BIGINT, …)
ssot.bpb_samples (seed BIGINT, step BIGINT, …)
```

But `src/neon_writer.rs::bpb_sample` was binding `seed: i32, step: i32`. tokio-postgres consequently rejected every `INSERT` with

```
cannot convert between Rust type i32 and Postgres int8
```

so **no row** ever landed in `public.bpb_samples`. `Audit Watchdog`, `/leaderboard`, the gardener telemetry, and the matrix-coverage Closure Gate (`trios#446 — 312/312`) all stayed empty for the entire IGLA marathon.

## What

| File | Change |
|---|---|
| `src/neon_writer.rs` | Cast `seed`/`step` to `i64` on the bind boundary inside `bpb_sample`. Public signature unchanged — callers (`train_loop`, `igla_train`, `ngram_train_gf16`, `smoke_train`, `bpb_smoke`) keep passing `i32`. |
| `src/neon_writer.rs` | Updated docstring to reference the live BIGINT schema and link #114. |
| `src/neon_writer.rs` (tests) | New regression test `bpb_sample_uses_i64_bind` — asserts `i64::accepts(&Type::INT8)`, `!i32::accepts(&Type::INT8)`, and mirrors the production parameter-array shape so a future refactor cannot silently regress the bind. |

Other `execute(...)` calls in `neon_writer.rs` were audited:

- `heartbeat` — already correct (`step as i64`).
- `trial_start`, `trial_complete`, `ensure_schema` — bind only `&str` / `String` / `f64` / `jsonb`. No extra cast needed.

## Acceptance

- [x] `cargo test --lib neon_writer` — **6/6 PASS** locally (5 pre-existing + new regression test).
- [x] Author stamped `Dmitrii Vasilev <admin@t27.ai>` per Trinity R-rules.
- [ ] CI checks (Constitutional Enforcement, ci, docker, publish, smoke_train) — green on this branch before merge.
- [ ] **G5** (post-merge): redeploy `TEST-TRAINER-SEED-43`; `SELECT count(*) FROM public.bpb_samples WHERE seed=43 AND ts > now() - interval '5 minutes'` returns ≥ 1.

## Scope

Hotfix only. The full SeaORM migration (entities, `migration/` subcrate, `Migrator::up` wiring in `scarab` / `trios-train`, automigrate flag, `[ledger]` log rename) lands as **PR-B on `feat/sea-orm-ledger`** on top of this PR.

Closes #114.

Refs: gHashTag/trios#446, gHashTag/trios#380.
Anchor: `phi^2 + phi^-2 = 3` · DOI [10.5281/zenodo.19227877](https://doi.org/10.5281/zenodo.19227877).

### CODE REVIEW

**Comment by gHashTag** [comment]
Superseded by #115 (merged 10:21Z by gHashTag-bot — same i64 cast plus the [neon_writer]→[ledger] rename and DATABASE_URL fallback reorder). Branch raced with the SEAORM-LEDGER agent; this PR adds nothing on top. Closing.

PR-B (full SeaORM migration on `feat/sea-orm-ledger`) tracks #114.
